File: BTFViewer/web/src/utils/aiEvidencePackage.js
/**
 * Compact AI evidence package (TODO Phase 6).
 * Keep in sync with btf_viewer_pkg/ai_evidence_package.py.
 *
 * Assemble only the evidence needed to answer one question — question + range,
 * trace-health status + limitations, related entities, required statistics with
 * units and sample counts, a short event context, and the picked findings /
 * bookmarks — plus stable ids and an explicit fact / inference / missing
 * instruction. Pure + deterministic; `estimateTokens` sizes it before sending
 * and `formatEvidencePackagePreview` renders exactly what goes out.
 */

export const EVIDENCE_PACKAGE_SCHEMA = 'btf-viewer-evidence/1'

export const RESPONSE_CONTRACT = [
  'Confirmed observations',
  'Possible explanations',
  'Contradicting or missing evidence',
  'Recommended verification steps',
  'Evidence references',
]

const INSTRUCTIONS =
  'Answer only from the evidence below. Separate confirmed observations from '
  + 'inferences, and name missing evidence explicitly. Every claim must cite an '
  + 'evidence id or a time/value from this package; mark any uncited statement '
  + 'as unverified. If the evidence is insufficient, say so.'

const SUMMARY_KEYS = [
  'span_ns', 'tasks', 'segments', 'sti_events', 'context_switches',
  'gap_avg_ns', 'gap_max_ns', 'migrations', 'migrated_tasks',
  'load_balance_score', 'load_balance_sigma', 'tick_health', 'missed_ticks',
  'time_scale',
]

function compactSummary(summary) {
  if (!summary || typeof summary !== 'object') return {}
  const out = {}
  for (const k of SUMMARY_KEYS) {
    if (summary[k] != null) out[k] = summary[k]
  }
  return out
}

function makeRedactor(redact) {
  const map = new Map()
  return (name) => {
    const s = String(name || '')
    if (!redact || !s) return s
    if (!map.has(s)) map.set(s, `task_${map.size + 1}`)
    return map.get(s)
  }
}

export function buildEvidencePackage({
  question = '', scope = '', analysisRange = null, traceName = '',
  traceSummary = null, health = null, findings = null, investigation = null,
  entities = null, cores = null, statistics = null, eventContext = null,
  redactNames = false,
} = {}) {
  const red = makeRedactor(redactNames)
  const inv = (investigation && typeof investigation === 'object') ? investigation : {}

  const fnd = (findings || []).filter(f => f && typeof f === 'object').map((f, i) => ({
    id: `F${i + 1}`,
    rule_id: String(f.rule_id || f.ruleId || f.id || ''),
    severity: String(f.severity || 'info'),
    observation: String(f.observation || f.title || f.text || ''),
    entities: (f.entities || []).map(red),
    measured_values: [...(f.measured_values || f.measuredValues || [])],
    comparison_basis: String(f.comparison_basis || f.comparisonBasis || ''),
    limitations: [...(f.limitations || [])],
  }))

  const bms = (inv.bookmarks || []).filter(b => b && typeof b === 'object').map(b => ({
    id: String(b.id || ''),
    type: String(b.type || 'observation'),
    title: String(b.title || ''),
    note: String(b.note || ''),
    refs: (b.refs || []).filter(r => r && typeof r === 'object').map(r => ({ ...r })),
  }))

  const stats = (statistics || []).filter(s => s && typeof s === 'object').map(s => ({
    name: String(s.name || ''),
    value: s.value ?? null,
    unit: String(s.unit || ''),
    sample_count: s.sample_count ?? s.sampleCount ?? null,
    entity: red(s.entity || ''),
  }))

  const evc = (eventContext || []).filter(e => e && typeof e === 'object').map(e => ({
    time: e.time ?? null,
    kind: String(e.kind || ''),
    detail: red(e.detail || e.task || ''),
    core: String(e.core || ''),
  }))

  let healthBlock = null
  if (health && typeof health === 'object') {
    healthBlock = {
      status: String(health.status || ''),
      issue_count: Number(health.issue_count ?? health.issueCount ?? 0) || 0,
      limitations: [...(health.metric_limitations || health.metricLimitations || [])],
    }
  }

  const pkg = {
    schema: EVIDENCE_PACKAGE_SCHEMA,
    question: String(question || '').trim(),
    scope: String(scope || '').trim(),
    analysis_range: (analysisRange && analysisRange.start != null)
      ? { start: Math.trunc(analysisRange.start), end: Math.trunc(analysisRange.end) }
      : null,
    trace: {
      name: redactNames ? '(redacted)' : String(traceName || ''),
      summary: compactSummary(traceSummary),
    },
    trace_health: healthBlock,
    entities: (entities || []).map(red),
    cores: (cores || []).map(String),
    statistics: stats,
    event_context: evc,
    findings: fnd,
    bookmarks: bms,
    conclusion_so_far: String(inv.conclusion || ''),
    unresolved_questions: (inv.unresolved_questions || []).map(String),
    instructions: INSTRUCTIONS,
    response_contract: [...RESPONSE_CONTRACT],
    redacted: !!redactNames,
  }
  if (redactNames) pkg.alias_note = 'Task names replaced with task_N aliases for privacy.'
  return pkg
}

export function estimateTokens(pkg) {
  const text = JSON.stringify(sortKeys(pkg))
  return Math.max(1, Math.floor((text.length + 3) / 4))
}

export function evidencePackageSize(pkg) {
  const text = JSON.stringify(sortKeys(pkg))
  return {
    bytes: new TextEncoder().encode(text).length,
    chars: text.length,
    approx_tokens: estimateTokens(pkg),
    findings: (pkg.findings || []).length,
    bookmarks: (pkg.bookmarks || []).length,
    statistics: (pkg.statistics || []).length,
    event_context: (pkg.event_context || []).length,
  }
}

export function formatEvidencePackagePreview(pkg) {
  const size = evidencePackageSize(pkg)
  const lines = []
  lines.push(`Evidence package — ~${size.approx_tokens} tokens, `
    + `${size.bytes.toLocaleString()} bytes`
    + (pkg.redacted ? '  (task names redacted)' : ''))
  lines.push('')
  lines.push(`Question: ${pkg.question || '(none)'}`)
  if (pkg.scope) lines.push(`Scope: ${pkg.scope}`)
  if (pkg.analysis_range) lines.push(`Range: ${pkg.analysis_range.start} – ${pkg.analysis_range.end}`)
  if (pkg.trace_health) {
    const lim = pkg.trace_health.limitations?.length
      ? `; limited: ${pkg.trace_health.limitations.join(', ')}` : ''
    lines.push(`Trace health: ${pkg.trace_health.status} (${pkg.trace_health.issue_count} issue(s))${lim}`)
  }
  if (pkg.entities?.length) lines.push(`Entities: ${pkg.entities.slice(0, 20).join(', ')}`)
  if (pkg.statistics?.length) {
    lines.push(`Statistics (${pkg.statistics.length}):`)
    for (const s of pkg.statistics.slice(0, 20)) {
      const sc = s.sample_count != null ? `, n=${s.sample_count}` : ''
      const ent = s.entity ? ` [${s.entity}]` : ''
      lines.push(`  - ${s.name}${ent} = ${s.value} ${s.unit}${sc}`.trimEnd())
    }
  }
  if (pkg.findings?.length) {
    lines.push(`Findings (${pkg.findings.length}):`)
    for (const f of pkg.findings) lines.push(`  [${f.id}] ${f.severity}: ${f.observation}`)
  }
  if (pkg.bookmarks?.length) {
    lines.push(`Bookmarks (${pkg.bookmarks.length}):`)
    for (const b of pkg.bookmarks) lines.push(`  (${b.type}) ${b.title}`)
  }
  if (pkg.event_context?.length) {
    lines.push(`Event context: ${pkg.event_context.length} event(s) around the range`)
  }
  if (pkg.conclusion_so_far) lines.push(`Conclusion so far: ${pkg.conclusion_so_far}`)
  lines.push('')
  lines.push('Response contract: ' + (pkg.response_contract || []).join(' · '))
  lines.push(pkg.instructions || '')
  return lines.join('\n') + '\n'
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys)
  if (v && typeof v === 'object') {
    const out = {}
    for (const k of Object.keys(v).sort()) out[k] = sortKeys(v[k])
    return out
  }
  return v
}

File: BTFViewer/btf_viewer_pkg/ai_evidence_package.py
"""Compact AI evidence package.

Assemble only the evidence needed to answer one question — the user's question
and range, trace-health status + limitations, related entities, the required
statistics with units and sample counts, a short event context around the
range, and the findings / bookmarks the user picked — plus stable identifiers
and an explicit fact / inference / missing-evidence instruction.

Pure and deterministic. ``estimate_tokens`` gives a rough size before anything
is sent; ``format_evidence_package_preview`` renders exactly what would go out.
Keep in sync with ``web/src/utils/aiEvidencePackage.js``.
"""
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional, Sequence

EVIDENCE_PACKAGE_SCHEMA = "btf-viewer-evidence/1"

RESPONSE_CONTRACT = [
    "Confirmed observations",
    "Possible explanations",
    "Contradicting or missing evidence",
    "Recommended verification steps",
    "Evidence references",
]

_INSTRUCTIONS = (
    "Answer only from the evidence below. Separate confirmed observations from "
    "inferences, and name missing evidence explicitly. Every claim must cite an "
    "evidence id or a time/value from this package; mark any uncited statement "
    "as unverified. If the evidence is insufficient, say so."
)


def _redact_name(name: str, redact: bool, alias_map: Dict[str, str]) -> str:
    if not redact or not name:
        return name
    if name not in alias_map:
        alias_map[name] = f"task_{len(alias_map) + 1}"
    return alias_map[name]


def build_evidence_package(
    *,
    question: str,
    scope: str = "",
    analysis_range: Optional[Dict[str, int]] = None,
    trace_name: str = "",
    trace_summary: Optional[Dict[str, Any]] = None,
    health: Optional[Dict[str, Any]] = None,
    findings: Optional[Sequence[Dict[str, Any]]] = None,
    investigation: Optional[Dict[str, Any]] = None,
    entities: Optional[Sequence[str]] = None,
    cores: Optional[Sequence[str]] = None,
    statistics: Optional[Sequence[Dict[str, Any]]] = None,
    event_context: Optional[Sequence[Dict[str, Any]]] = None,
    redact_names: bool = False,
) -> Dict[str, Any]:
    """Build a compact, provider-independent evidence package dict."""
    alias_map: Dict[str, str] = {}

    def red(v: str) -> str:
        return _redact_name(str(v or ""), redact_names, alias_map)

    fnd = []
    for i, f in enumerate(findings or []):
        if not isinstance(f, dict):
            continue
        fnd.append({
            "id": f"F{i + 1}",
            "rule_id": str(f.get("rule_id") or f.get("id") or ""),
            "severity": str(f.get("severity") or "info"),
            "observation": str(f.get("observation") or f.get("title") or f.get("text") or ""),
            "entities": [red(e) for e in (f.get("entities") or [])],
            "measured_values": list(f.get("measured_values") or []),
            "comparison_basis": str(f.get("comparison_basis") or ""),
            "limitations": list(f.get("limitations") or []),
        })

    bms = []
    inv = investigation if isinstance(investigation, dict) else {}
    for b in inv.get("bookmarks", []) or []:
        if not isinstance(b, dict):
            continue
        bms.append({
            "id": str(b.get("id") or ""),
            "type": str(b.get("type") or "observation"),
            "title": str(b.get("title") or ""),
            "note": str(b.get("note") or ""),
            "refs": [dict(r) for r in (b.get("refs") or []) if isinstance(r, dict)],
        })

    stats = []
    for s in statistics or []:
        if not isinstance(s, dict):
            continue
        stats.append({
            "name": str(s.get("name") or ""),
            "value": s.get("value"),
            "unit": str(s.get("unit") or ""),
            "sample_count": s.get("sample_count"),
            "entity": red(s.get("entity") or ""),
        })

    evc = []
    for e in event_context or []:
        if not isinstance(e, dict):
            continue
        evc.append({
            "time": e.get("time"),
            "kind": str(e.get("kind") or ""),
            "detail": red(e.get("detail") or e.get("task") or ""),
            "core": str(e.get("core") or ""),
        })

    health_block = None
    if isinstance(health, dict):
        health_block = {
            "status": str(health.get("status") or ""),
            "issue_count": int(health.get("issue_count", health.get("issueCount", 0)) or 0),
            "limitations": list(
                health.get("metric_limitations") or health.get("metricLimitations") or []),
        }

    package: Dict[str, Any] = {
        "schema": EVIDENCE_PACKAGE_SCHEMA,
        "question": str(question or "").strip(),
        "scope": str(scope or "").strip(),
        "analysis_range": (
            {"start": int(analysis_range["start"]), "end": int(analysis_range["end"])}
            if isinstance(analysis_range, dict) and analysis_range.get("start") is not None
            else None
        ),
        "trace": {
            "name": "(redacted)" if redact_names else str(trace_name or ""),
            "summary": _compact_summary(trace_summary),
        },
        "trace_health": health_block,
        "entities": [red(e) for e in (entities or [])],
        "cores": [str(c) for c in (cores or [])],
        "statistics": stats,
        "event_context": evc,
        "findings": fnd,
        "bookmarks": bms,
        "conclusion_so_far": str(inv.get("conclusion") or ""),
        "unresolved_questions": [str(q) for q in (inv.get("unresolved_questions") or [])],
        "instructions": _INSTRUCTIONS,
        "response_contract": list(RESPONSE_CONTRACT),
        "redacted": bool(redact_names),
    }
    if redact_names and alias_map:
        package["alias_note"] = "Task names replaced with task_N aliases for privacy."
    return package


_SUMMARY_KEYS = (
    "span_ns", "tasks", "segments", "sti_events", "context_switches",
    "gap_avg_ns", "gap_max_ns", "migrations", "migrated_tasks",
    "load_balance_score", "load_balance_sigma", "tick_health", "missed_ticks",
    "time_scale",
)


def _compact_summary(summary: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if not isinstance(summary, dict):
        return {}
    return {k: summary[k] for k in _SUMMARY_KEYS if k in summary and summary[k] is not None}


# --------------------------------------------------------------------------
def estimate_tokens(package: Dict[str, Any]) -> int:
    """Rough token estimate for the serialised package (~4 chars/token)."""
    text = json.dumps(package, ensure_ascii=False, sort_keys=True)
    return max(1, (len(text) + 3) // 4)


def evidence_package_size(package: Dict[str, Any]) -> Dict[str, int]:
    text = json.dumps(package, ensure_ascii=False, sort_keys=True)
    return {
        "bytes": len(text.encode("utf-8")),
        "chars": len(text),
        "approx_tokens": estimate_tokens(package),
        "findings": len(package.get("findings") or []),
        "bookmarks": len(package.get("bookmarks") or []),
        "statistics": len(package.get("statistics") or []),
        "event_context": len(package.get("event_context") or []),
    }


def format_evidence_package_preview(package: Dict[str, Any]) -> str:
    """Human-readable preview of exactly what would be sent."""
    lines: List[str] = []
    size = evidence_package_size(package)
    lines.append(f"Evidence package — ~{size['approx_tokens']} tokens, "
                 f"{size['bytes']:,} bytes"
                 + ("  (task names redacted)" if package.get("redacted") else ""))
    lines.append("")
    lines.append(f"Question: {package.get('question') or '(none)'}")
    if package.get("scope"):
        lines.append(f"Scope: {package['scope']}")
    rng = package.get("analysis_range")
    if rng:
        lines.append(f"Range: {rng['start']} – {rng['end']}")
    th = package.get("trace_health")
    if th:
        lim = f"; limited: {', '.join(th['limitations'])}" if th.get("limitations") else ""
        lines.append(f"Trace health: {th['status']} ({th['issue_count']} issue(s)){lim}")
    if package.get("entities"):
        lines.append(f"Entities: {', '.join(package['entities'][:20])}")
    if package.get("statistics"):
        lines.append(f"Statistics ({len(package['statistics'])}):")
        for s in package["statistics"][:20]:
            sc = f", n={s['sample_count']}" if s.get("sample_count") is not None else ""
            ent = f" [{s['entity']}]" if s.get("entity") else ""
            lines.append(f"  - {s['name']}{ent} = {s['value']} {s['unit']}{sc}".rstrip())
    if package.get("findings"):
        lines.append(f"Findings ({len(package['findings'])}):")
        for f in package["findings"]:
            lines.append(f"  [{f['id']}] {f['severity']}: {f['observation']}")
    if package.get("bookmarks"):
        lines.append(f"Bookmarks ({len(package['bookmarks'])}):")
        for b in package["bookmarks"]:
            lines.append(f"  ({b['type']}) {b['title']}")
    if package.get("event_context"):
        lines.append(f"Event context: {len(package['event_context'])} event(s) "
                     "around the range")
    if package.get("conclusion_so_far"):
        lines.append(f"Conclusion so far: {package['conclusion_so_far']}")
    lines.append("")
    lines.append("Response contract: " + " · ".join(package.get("response_contract", [])))
    lines.append(package.get("instructions", ""))
    return "\n".join(lines) + "\n"
